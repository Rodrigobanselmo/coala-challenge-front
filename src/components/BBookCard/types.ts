import { IBook } from "@/core/interfaces/IBook";
import { CircularProgressProps } from "@mui/material/CircularProgress";
import { IconButtonProps } from "@mui/material/IconButton";

export type BBookCardProps = {
  book: IBook;
  buttonLabel?: string;
  onClickButton?: (book: IBook) => Promise<void>;
};
