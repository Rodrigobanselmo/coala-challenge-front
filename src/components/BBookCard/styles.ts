import { Box, Button, IconButton, Typography, styled } from "@mui/material";

export const StyledGridContainer = styled(Box)`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  grid-gap: 20px;
  margin-bottom: 20px;
  width: 100%;
`;

export const StyledCard = styled(Box)`
  border-radius: 16px;
  text-align: center;
  transition: transform 0.2s ease-in-out;
  position: relative;
  min-height: 400px;
  margin: 5px;
  min-width: 200px;
  position: relative;
  overflow: hidden;
  padding-bottom: 50px;

  &:hover {
    transform: scale(1.05);
    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  }
`;

export const StyledTitle = styled("h2")`
  margin: 0;
  font-size: 1rem;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: normal;
`;

export const StyledAuthors = styled(Typography)`
  margin-top: 8px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const StyledExchangeButton = styled(Button)<{ is_visible: number }>`
  border: none;
  border-radius: 4px;
  padding: 8px 16px;
  cursor: pointer;
  transition: background-color 0.3s, opacity 0.3s;
  margin-top: 16px;
  opacity: ${(props) => (props.is_visible ? "1" : "0")};
  position: absolute;
  bottom: 16px;
  left: 50%;
  width: 80%;
  transform: translateX(-50%);
  font-size: 0.8rem;
`;

export const StyledBackgroundImage = styled(Box)`
  border: 1px solid;
  border-color: ${(props) => props.theme.palette.divider};
  border-radius: 16px;
  width: 100%;
  height: 300px;
  overflow: hidden;
  position: relative;
`;

export const StyledInfoContainer = styled(Box)`
  width: 100%;
  padding: 16px;
`;
