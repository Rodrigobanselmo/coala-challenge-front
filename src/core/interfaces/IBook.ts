export type IBook = {
  id: number;
  title: string;
  description: string;
  authors?: string[];
  categories?: string[];
  pageCount?: number;
  googleId?: string;
  publishedAt?: Date;
  language?: string;
  smallThumbnail?: string;
  thumbnail?: string;
  updatedAt: Date;
  createdAt: Date;
  deletedAt?: Date;
};
