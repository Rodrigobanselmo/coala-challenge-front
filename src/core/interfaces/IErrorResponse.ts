import { IconProps } from "@mui/material";

export interface IErrorResponse {
  response: {
    status: number;
    data: {
      message: string;
    };
  };
}
