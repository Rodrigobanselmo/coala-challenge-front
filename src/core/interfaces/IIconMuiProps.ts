import { IconProps } from "@mui/material";

export interface IIconMuiProps {
  muiName?: string;
  fontSize?: IconProps["fontSize"];
  color?: IconProps["color"];
}
