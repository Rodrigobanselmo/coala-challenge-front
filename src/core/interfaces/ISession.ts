import { IUser } from "./IUser";

export type ISession = {
  token: string;
  user: IUser;
};
