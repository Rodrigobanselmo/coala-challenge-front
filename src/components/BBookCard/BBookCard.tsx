import { IBook } from "@/core/interfaces/IBook";
import {
  StyledAuthors,
  StyledBackgroundImage,
  StyledExchangeButton,
  StyledCard,
  StyledInfoContainer,
  StyledTitle,
} from "./styles";
import Image from "next/image";
import { useState } from "react";
import { BBookCardProps } from "./types";

export const BBookCard = ({
  book,
  buttonLabel,
  onClickButton,
}: BBookCardProps) => {
  const [isButtonVisible, setIsButtonVisible] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  const handleClickButton = async () => {
    if (!onClickButton || isLoading) return;
    setIsLoading(true);
    onClickButton(book).finally(() => setIsLoading(false));
  };

  return (
    <StyledCard
      onMouseEnter={() => setIsButtonVisible(true)}
      onMouseLeave={() => setIsButtonVisible(false)}
    >
      <StyledBackgroundImage>
        <Image
          src={book.thumbnail || "/images/book-placeholder.png"}
          alt={book.title}
          fill
          style={{ objectFit: "cover" }}
        />
      </StyledBackgroundImage>
      <StyledInfoContainer>
        <StyledTitle>{book.title}</StyledTitle>
        <StyledAuthors>{book.authors?.join(", ")}</StyledAuthors>
        {onClickButton && (
          <StyledExchangeButton
            onClick={handleClickButton}
            is_visible={isButtonVisible}
            variant="contained"
            color="secondary"
            disabled={isLoading}
          >
            {isLoading ? "carregando..." : buttonLabel}
          </StyledExchangeButton>
        )}
      </StyledInfoContainer>
    </StyledCard>
  );
};
