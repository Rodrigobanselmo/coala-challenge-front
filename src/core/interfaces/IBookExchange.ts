import { StatusEnum } from "../enums/status.enum";
import { IBook } from "./IBook";
import { IUser } from "./IUser";

export type IBookExchange = {
  id: number;
  interestBookId: number;
  targetUserId: string;
  askingBookId: number;
  interetUserId: string;
  updatedAt: Date;
  createdAt: Date;
  deletedAt: Date | null;
  interestBook?: IBook;
  targetUser?: IUser;
  askingBook?: IBook;
  interetUser?: IUser;
  status: StatusEnum;
};
