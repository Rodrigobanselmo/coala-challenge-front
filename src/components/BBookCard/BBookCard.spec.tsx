import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { BBookCard } from "./BBookCard";
import { IBook } from "@/core/interfaces/IBook";

// Mocked book data for testing
const mockBook: IBook = {
  title: "Test Book",
  authors: ["Author 1", "Author 2"],
  thumbnail: "/test-image.png",
  id: 1,
  createdAt: new Date(),
  updatedAt: new Date(),
};

describe("BBookCard", () => {
  const mockButtonLabel = "Test Button";
  const mockOnClickButton = jest.fn();

  it("matches the snapshot", () => {
    const { container } = render(
      <BBookCard book={mockBook} buttonLabel="Test Button" />
    );

    expect(container).toMatchSnapshot();
  });

  it("matches the snapshot with button", () => {
    const { container } = render(
      <BBookCard
        book={mockBook}
        buttonLabel="Test Button"
        onClickButton={async () => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders the book card with provided data", () => {
    render(
      <BBookCard
        book={mockBook}
        buttonLabel={mockButtonLabel}
        onClickButton={mockOnClickButton}
      />
    );

    const titleElement = screen.getByText(mockBook.title);
    const authorsElement = screen.getByText(mockBook.authors?.join(", ") || "");
    const buttonElement = screen.getByText(mockButtonLabel);
    const imageElement = screen.getByAltText(mockBook.title);

    expect(titleElement).toBeInTheDocument();
    expect(authorsElement).toBeInTheDocument();
    expect(buttonElement).toBeInTheDocument();
    expect(imageElement).toBeInTheDocument();
  });

  it("calls onClickButton when the button is clicked", async () => {
    render(
      <BBookCard
        book={mockBook}
        buttonLabel={mockButtonLabel}
        onClickButton={mockOnClickButton}
      />
    );

    const buttonElement = screen.getByText(mockButtonLabel);

    fireEvent.click(buttonElement);

    expect(mockOnClickButton).toHaveBeenCalledWith(mockBook);

    await waitFor(() => {
      expect(screen.getByText("carregando...")).toBeInTheDocument();
    });
  });

  it("shows the button on mouse hover", () => {
    render(
      <BBookCard
        book={mockBook}
        buttonLabel={mockButtonLabel}
        onClickButton={mockOnClickButton}
      />
    );

    const cardElement = screen.getByTestId("StyledCard");

    fireEvent.mouseEnter(cardElement);

    expect(screen.getByText(mockButtonLabel)).toBeVisible();
  });

  it("hides the button on mouse leave", () => {
    render(<BBookCard book={mockBook} buttonLabel={mockButtonLabel} />);

    const cardElement = screen.getByTestId("StyledCard");

    fireEvent.mouseEnter(cardElement);
    fireEvent.mouseLeave(cardElement);

    expect(screen.queryByText(mockButtonLabel)).not.toBeInTheDocument();
  });
});
