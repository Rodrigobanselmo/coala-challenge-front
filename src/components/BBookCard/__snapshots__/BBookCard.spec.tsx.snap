// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BBookCard matches the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-xtnipu"
    data-testid="StyledCard"
  >
    <div
      class="MuiBox-root css-rrl90c"
    >
      <img
        alt="Test Book"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Ftest-image.png&w=3840&q=75"
        srcset="/_next/image?url=%2Ftest-image.png&w=640&q=75 640w, /_next/image?url=%2Ftest-image.png&w=750&q=75 750w, /_next/image?url=%2Ftest-image.png&w=828&q=75 828w, /_next/image?url=%2Ftest-image.png&w=1080&q=75 1080w, /_next/image?url=%2Ftest-image.png&w=1200&q=75 1200w, /_next/image?url=%2Ftest-image.png&w=1920&q=75 1920w, /_next/image?url=%2Ftest-image.png&w=2048&q=75 2048w, /_next/image?url=%2Ftest-image.png&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
      />
    </div>
    <div
      class="MuiBox-root css-mjw23k"
    >
      <h2
        class="css-yykxhl"
      >
        Test Book
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body1 css-17hjpk6-MuiTypography-root"
      >
        Author 1, Author 2
      </p>
    </div>
  </div>
</div>
`;

exports[`BBookCard matches the snapshot with button 1`] = `
<div>
  <div
    class="MuiBox-root css-xtnipu"
    data-testid="StyledCard"
  >
    <div
      class="MuiBox-root css-rrl90c"
    >
      <img
        alt="Test Book"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Ftest-image.png&w=3840&q=75"
        srcset="/_next/image?url=%2Ftest-image.png&w=640&q=75 640w, /_next/image?url=%2Ftest-image.png&w=750&q=75 750w, /_next/image?url=%2Ftest-image.png&w=828&q=75 828w, /_next/image?url=%2Ftest-image.png&w=1080&q=75 1080w, /_next/image?url=%2Ftest-image.png&w=1200&q=75 1200w, /_next/image?url=%2Ftest-image.png&w=1920&q=75 1920w, /_next/image?url=%2Ftest-image.png&w=2048&q=75 2048w, /_next/image?url=%2Ftest-image.png&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
      />
    </div>
    <div
      class="MuiBox-root css-mjw23k"
    >
      <h2
        class="css-yykxhl"
      >
        Test Book
      </h2>
      <p
        class="MuiTypography-root MuiTypography-body1 css-17hjpk6-MuiTypography-root"
      >
        Author 1, Author 2
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1g5ile9-MuiButtonBase-root-MuiButton-root"
        is_visible="0"
        tabindex="0"
        type="button"
      >
        Test Button
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
